Add support for EasyCLA

anusha94 opened the pull request:
Added a new field `EasyCLAEnabled` in the plugins config and based on
this field, if set will assign the `claContextName` to `cla/easy-cla`, else
default to `cla/linuxfoundation`

Fixes #22721

anusha94 commented:
/cc @nikhita

anusha94 commented:
Fixes #22742

MadhavJivrajani commented:
/ok-to-test

MadhavJivrajani commented on prow/plugins/cla/cla.go:
```diff
@@ -58,12 +57,19 @@ It may take a couple minutes for the CLA signature to be fully registered; after
 func init() {
 	plugins.RegisterStatusEventHandler(pluginName, handleStatusEvent, helpProvider)
 	plugins.RegisterGenericCommentHandler(pluginName, handleCommentEvent, helpProvider)
+	config := &plugins.Configuration{}
+	if config.EasyCLAEnabled {
+		claContextName = "cla/easy-cla"
+	} else {
+		claContextName = "cla/linuxfoundation"
+	}
```
If I'm not wrong, this will always set the `claContextName` to `cla/linuxfoundation` because `EasyCLAEnabled` will default to `false` in an empty `Configuration` struct?

anusha94 replied:
Ahhh...right. Thanks!

MadhavJivrajani commented:
Fixes #22721

MadhavJivrajani commented:
/label tide/merge-method-squash

cjwagner commented on prow/plugins/cla/cla_test.go:
```diff
@@ -140,6 +141,21 @@ func TestCLALabels(t *testing.T) {
 			removedLabels: []string{fmt.Sprintf("/#3:%s", labels.ClaYes)},
 		},
+		{
+			name:           "cla/linuxfoundation status failure removes \"cncf-cla: yes\" label",
```
Please add test cases for successful status cases as well.

cjwagner reviewed (commented):
Thanks for the PR!

anusha94 commented on prow/plugins/config.go:
```diff
@@ -90,6 +90,7 @@ type Configuration struct {
 	Welcome              []Welcome                    `json:"welcome,omitempty"`
 	Override             Override                     `json:"override,omitempty"`
 	Help                 Help                         `json:"help,omitempty"`
+	EasyCLAEnabled       bool                         `json:"easy_cla_enabled,omitempty"`
```
Got it.

anusha94 commented:
@cjwagner Thanks for the review :)

anusha94 commented:
/retest

MadhavJivrajani commented:
@anusha94 seems like you have to run these commands to update the bazel rules and test fixtures:
```bash
# update hazel rules
bazel run @io_k8s_repo_infra//hack:update-bazel
# update dependencies
bazel run @io_k8s_repo_infra//hack:update-deps
# update test fixtures based on your changes (addition of `CLAConfig` struct)
UPDATE=true go test ./prow/plugins/... -run TestGenYamlDocs
```

anusha94 commented:
Thanks @MadhavJivrajani  :)

MadhavJivrajani commented:
/cc @mrbobbytables

MadhavJivrajani commented on prow/plugins/cla/cla.go:
```diff
@@ -93,20 +92,20 @@ type gitHubClient interface {
+	return handle(pc.GitHubClient, pc.Logger, se, pc.PluginConfig.CLAConfig)
 }
 
 // 1. Check that the status event received from the webhook is for the CNCF-CLA.
 // 2. Use the github search API to search for the PRs which match the commit hash corresponding to the status event.
 // 3. For each issue that matches, check that the PR's HEAD commit hash against the commit hash for which the status
 //    was received. This is because we only care about the status associated with the last (latest) commit in a PR.
 // 4. Set the corresponding CLA label if needed.
```
nit: could you update this comment to reflect the new cla logic? I think only point `1` would need updating

cjwagner commented on prow/plugins/cla/cla.go:
```diff
@@ -242,7 +241,7 @@ func handleComment(gc gitHubClient, log *logrus.Entry, e *github.GenericCommentE
 	for _, status := range combined.Statuses {
 
 		// Only consider "cla/linuxfoundation" status.
```
This comment is out of date.

cjwagner commented on prow/plugins/config.go:
```diff
@@ -482,6 +483,11 @@ type ConfigUpdater struct {
 
+// CLAConfig contains the list of available context names
+type CLAConfig struct {
+	CLAContextNames []string `json:"cla_context_names,omitempty"`
```
Without defaulting logic this is a breaking change for all existing users. We need to default `CLAContextNames` to `[]string{"cla/linuxfoundation"}` if it is not specified. (Since the empty slice would never be a reasonable config we can just check if the slice is empty to determine if we should default.)
Please add defaulting logic to this function: https://github.com/kubernetes/test-infra/blob/13e6fe9adce3cc1929cd96ad5816f20c79306b5b/prow/plugins/config.go#L947-L994

cjwagner reviewed (changes requested):
I see one breaking change that we need to avoid, but otherwise looks good to me.

mrbobbytables commented on prow/plugins/cla/cla.go:
```diff
@@ -93,15 +93,21 @@ type gitHubClient interface {
+	if isEasyCLAEnabled {
+		claContextName = "cla/easy-cla"
+	} else {
+		claContextName = "cla/linuxfoundation"
```
> When migrating from the current CLA service to EasyCLA there will likely be some PRs with the old status context, some with the new, and maybe even some with both. A single boolean toggle to switch the context we respect doesn't allow us to handle this case gracefully. (@mrbobbytables appears to be more familiar with this and can hopefully confirm if my assumptions about the migration process are true.)

💯 

We want to be able to support both in the event there is an issue with easycla and we have to stay on the current version.

anusha94 commented:
@cjwagner Could you please review again?

jeefy commented:
/lgtm

spiffxp commented on prow/plugins/cla/cla.go:
```diff
@@ -241,8 +240,8 @@ func handleComment(gc gitHubClient, log *logrus.Entry, e *github.GenericCommentE
 
-		// Only consider "cla/linuxfoundation" status.
-		if status.Context == claContextName {
+		// Valid status values are cla/linuxfoundation and EasyCLA
+		if contains(cc.CLAContextNames, status.Context) {
```
Can we rewrite this so that it iterates through statuses to determine CLA yes/no first, and then once outside the for loop, applies labels accordingly?  I'd like to avoid wasting GitHub tokens on label apply/remove calls while we decide which status wins if they start competing.

spiffxp replied:
Is `combined.Statuses` sorted at this point? or could we be walking these in random order?

anusha94 replied:
@spiffxp Can you please elaborate on competing statuses? Going by the existing code `if status.Context == claContextName`, I assumed the `for loop` should not continue if it is not one of `cla/linuxfoundation` or `EasyCLA`

Or are you saying this entire `for loop `should be rewritten such that applying of `ClaNo` or `ClaYes` should be outside the for loop?
Thanks.

anusha94 replied:
> Is combined.Statuses sorted at this point? or could we be walking these in random order?

@mrbobbytables Do you know about the sort order of `combined.Statuses`? Or rather, how can I check / verify the order?

spiffxp replied:
>Can you please elaborate on competing statuses?

I missed the `break` down below.  You're right, one of them will win.  But will it be the same one every time?  And how will I know which one that is?

anusha94 replied:
I removed the `break` statement so that all contexts will be processed.

spiffxp commented on prow/plugins/plugin-config-documented.yaml:
```diff
@@ -316,6 +316,9 @@ cherry_pick_unapproved:
     # `do-not-merge/cherry-pick-not-approved` label.
     comment: ' '
+cla_config:
+    cla_context_names:
```
Please regenerate docs so this field has a useful comment

spiffxp commented on prow/plugins/cla/cla_test.go:
```diff
@@ -140,6 +141,34 @@ func TestCLALabels(t *testing.T) {
 		},
+		{
+			name:      "EasyCLA status failure removes \"cncf-cla: yes\" label",
+			context:   "EasyCLA",
+			state:     "failure",
```
Since the plugin now supports multiple contexts, I think it's fair to ask the test cases be written to test multiple contexts (or that you add a new Test that supports this).

I would like to see test cases with multiple contexts having conflicting values, so we can confirm that it's last-context-wins in all cases

spiffxp replied:
e.g. I would like to see something like
```go
contexts:
  "cla/linuxfoundation": "success" 
  "EasyCLA": "failure"
```

and vice-versa

spiffxp reviewed (changes requested):
I have concerns over dueling contexts

spiffxp commented on prow/plugins/cla/cla.go:
```diff
@@ -93,20 +92,20 @@ type gitHubClient interface {
 }
 
-// 1. Check that the status event received from the webhook is for the CNCF-CLA.
+// 1. Check that the status event received from the webhook is for the CNCF-CLA. The valid values are cla/linuxfoundation and EasyCLA.
```
If these are the only valid values, I would add validation logic confirming as much.  Otherwise, this is a configuration-specific detail

anusha94 replied:
@mrbobbytables 
`The valid values are cla/linuxfoundation and EasyCLA` Is this a right statement to make?

spiffxp replied:
The comment "the valid values..." does not match the state of this code.  You're not validating that these are the only two values allowed.  _Any_ values are allowed as configuration data.  This plugin could just as easily be configured to apply the CLA label if the the `monkeys` status context was successful.  I think this is great.  Generally, more configurability for re-use is good.

If you agree with this, then you should drop the "the valid values" part of the comment.  But, If you disagree, and think this should be purpose-built for just those two status contexts, then you should validate that these are the only valid values in some kind of `validate` function.

anusha94 replied:
Okay I get it now. I am going to remove the comment that those are the only two valid values.

spiffxp commented on prow/plugins/cla/cla_test.go:
```diff
@@ -140,6 +141,34 @@ func TestCLALabels(t *testing.T) {
 		},
+		{
+			name:      "EasyCLA status failure removes \"cncf-cla: yes\" label",
+			context:   "EasyCLA",
```
More of a nice-to-have cosmetic thing... is `EasyCLA` hardcoded on your end? It would be preferable to have something like `cla/EasyCLA`

mrbobbytables replied:
On the LF side, its hard-coded =/
Here is an example:
https://api.github.com/repos/CNCF-EasyCLAV2-Test/Repository1/statuses/31eb937fb47d258b060a762165f293de6f99d191

anusha94 replied:
Just to confirm - so this remains as `EasyCLA` right? And not `cla/EasyCLA`

spiffxp replied:
Sure, if it's hardcoded I guess it has to be `EasyCLA`

anusha94 commented:
@spiffxp 

I removed the `break` statement from [here](https://github.com/kubernetes/test-infra/pull/22742/files#diff-45486175573234cb728c8f2bfc7f2da46d7431b8eb4f3f049bbbcc73c119af15L283) in order for all the contexts to be processed.

In `TestCheckCLA`, I changed the [context field from string to a map](https://github.com/kubernetes/test-infra/pull/22742/files#diff-1e758035e054060499e235849f012fda0bfbd157353c30ced428f790e0c38994R214). Does this seem okay? Or having a separate `Test` for multiple contexts make more sense?

spiffxp commented on prow/plugins/cla/cla.go:
```diff
@@ -278,10 +276,18 @@ func handleComment(gc gitHubClient, log *logrus.Entry, e *github.GenericCommentE
 					}
 				}
 			}
-
-			// No need to consider other contexts once you find the one you need.
-			break
```
Since there's no break anymore, now we _are_ potentially adding/removing labels twice.

What I asked for in last review was to make the decision before acting, e.g.
```golang
claState := github.StatusPending
for _, status := range combined.Statuses {
  // code that sets claState
}
// code that sets labels based on claState
```

spiffxp commented on prow/plugins/cla/cla_test.go:
```diff
@@ -348,11 +425,20 @@ func TestCheckCLA(t *testing.T) {
+			}
+			var statuses []github.Status
+			for context, state := range tc.contexts {
+				statuses = append(statuses, github.Status{
```
re: my "are the statuses sorted" comment above, your unit tests are relying on _this_ being the sort

what I lack is proof that you've verified this sort is what you actually see in production (show me code that presorts the statuses, or docs from github guaranteeing the order in which they deliver statuses)

anusha94 replied:
@spiffxp 

I hear you. But I could not find proof for sort order. Here is the docs for [getting combined status](https://docs.github.com/en/rest/reference/repos#get-the-combined-status-for-a-specific-reference)
and the [client code to get the same](https://github.com/google/go-github/blob/ac48d59eb13be558759e9dd1e6f8f75126a83a4e/github/repos_statuses.go#L114)

spiffxp replied:
maybe this is a sign you should sort them before iterating over them?  Or, flip the loop so you iterate over the context names that you care about in the order you care about, and then see if they're contained in combined statuses

spiffxp commented on prow/plugins/cla/cla_test.go:
```diff
@@ -161,7 +190,10 @@ func TestCLALabels(t *testing.T) {
+		cc := plugins.CLAConfig{
+			CLAContextNames: []string{"cla/linuxfoundation", "EasyCLA"},
+		}
+		if err := handle(fc, logrus.WithField("plugin", pluginName), se, cc); err != nil {
```
OK, I think I assumed there was less duplication of logic before

So the way I read the StatusEvent handling code, is that it will respond to whatever the status is.  So if the CLA plugin is configured to listen to two CLA statuses, we will have a race between which of the two statuses reports in first.  Slowest status wins.

I think what would provide more consistency is to try to de-dupe the logic and do what `handleComment` does: get the PR's combined statuses, and ensure one of them consistently wins the race.  

e.g. let's say cla/linuxfoundation consistently wins:
```yaml
# happy path
state: 
  cla/linuxfoundation: pending
  EasyCLA: pending
  labels: []
receive:
  cla/linuxfoundation: success
result:
  add: "cncf-cla: yes"

# conflict: EasyCLA arrives last
state: 
  cla/linuxfoundation: success
  EasyCLA: pending
  labels: ["cncf-cla: yes"]
receive:
  EasyCLA: failure
result:
  # do nothing

# conflict: cla/linuxfoundation arrives last
state: 
  cla/linuxfoundation: pending
  EasyCLA: success
  labels: ["cncf-cla: yes"]
receive:
  cla/linuxfoundation: failure
result:
  add: "cncf-cla: no"
  remove: "cncf-cla: yes"
```

anusha94 replied:
>  So if the CLA plugin is configured to listen to two CLA statuses

From the docs, `The most recent status for each context is returned`.
I think my question is - are `EasyCLA` and `cla/linuxfoundation` considered as two different contexts or different names for a single `CLA` context? How does getting combined status here work? Combined status is still going to return both the contexts - how do we know which one is the latest?

spiffxp replied:
>I think my question is - are EasyCLA and cla/linuxfoundation considered as two different contexts or different names for a single CLA context

Well, I'm assuming I'm going to see two entries, one called `cla/linuxfoundation` and one called `EasyCLA` while the CNCF has both things running simultaneously on their end.  So they're different contexts.  You're updating the plugin to decide which one to treat as the authoritative source of truth from the CNCF.  This allows us to have the plugin pay attention to the new EasyCLA check, but defer to the old cla/linuxfoundation check as authoritative, without having to change anything.

Another possibly simpler option is to roll back some of what you've done, and update the plugin to listen to one-and-only-one context.  In this scenario we would:
- make the plugin listen to a configurable context
- start with cla/linuxfoundation as this plugin's source of truth
- setup EasyCLA on the CNCF's end to start reporting to PRs
- setup prow/branch-protector to treat EasyCLA contexts as non-blocking
- wait
- generate a report that shows over time, for all PRs within a given window, PRs where cla/linuxfoundation and EasyCLA differed
- figure out why and fix
- wait again
- generate report again
- repeat until there is no difference, and then we change branch-protector to make the EasyCLA context required, the cla/linuxfoundation context non-blocking, and the cla plugin to listen to EasyCLA to apply "cncf-cla: yes" labels

spiffxp replied:
>How does getting combined status here work? Combined status is still going to return both the contexts - how do we know which one is the latest?

What I was saying above is that the `handle` function _doesn't call GetCombinedStatus_, it only consider the status from the StatusEvent it's handling.  So it will get called once whenever `cla/linuxfoundation` is update, and again whenever `EasyCLA` is updated, and there's no way to guarantee which of those is going to arrive first.

Calling GetCombinedStatus to get both contexts is the point.  We don't _care_ which one came in latest, we only care that it came in for the commit we're considering, and that the authoritative context (if set) always overrides the others.

Again, I'm thinking this may all be too confusing, and it might be simpler to support a single (configurable) context only.  But we need to see how EasyCLA behaves live (in terms of responsiveness, reliability, consistency, and agreement with cla/linuxfoundation) before I will ever consider changing our workflow to pay attention to that over cla/linuxfoundation.

mrbobbytables replied:
EasyCLA has been enabled for the [k/org](https://github.com/kubernetes/org) repo in a non-blocking way as a test.
For some additional information, EasyCLA is currently in use by 3 other CNCF projects:
- Open telemetry (47 repos)
- grpc (26 repos)
- cloud custodian (6 repos)

For gRPC it's been some time (most of the kinks in the project were worked out with this one), I don't know when it was enabled for Otel or cloud custodian.

The LF has said if we do have problems we can roll back.

spiffxp reviewed (changes requested):
I'm not confident enough to deploy this plugin in a manner that listens to two status contexts simultaneously yet.  I wrote down some test cases I would want to see exercised to ensure consistency vs. a race between last-context-reporting-wins.

I can be convinced to allow the EasyCLA check to report to PRs as non-blocking, without this plugin configured to listen to it, to allow us to gather data on how often we encounter conflict between the old CLA check and the new CLA check.

spiffxp commented on prow/plugins/cla/cla_test.go:
```diff
@@ -332,6 +373,42 @@ func TestCheckCLA(t *testing.T) {
 		},
+		{
+			name: "cla/linuxfoundation status retains the cla-no label and removes cla-yes label when its state is \"failure\"",
+			contexts: map[string]string{
+				"cla/linuxfoundation": "failure",
+				"EasyCLA":             "success",
+			},
```
Thanks for the refactor down here.  I think I'm asking for a similar refactor to the status event handling test

anusha94 replied:
sure. will do :)

anusha94 commented:
/hold

MadhavJivrajani commented:
FYI https://github.com/kubernetes/test-infra/pull/23648
cc @spiffxp @anusha94

MadhavJivrajani commented:
/remove-lifecycle stale

MadhavJivrajani commented:
/close
closing this out in favor of a simpler switch over. Will send in a PR soon!
